### PR TITLE
New scheduler integration

### PR DESCRIPTION
**Pull request by istoica**
MVP support for the new scheduler for direct task calls. It does not work for actors yet, since actors are not created using direct task calls code path. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by ericl** [comment]
Somehow this PR is showing 254 changed files, could you fix this? Maybe master needs to be pulled in.

File: python/ray/tests/test_mini.py [hunk 1]
```diff
@@ -16,28 +16,32 @@ def f_simple():
...
+#    @ray.remote
+#    def f_args_by_value(x):
+#        return x
+#
+#    for arg in test_values:
+#        assert ray.get(f_args_by_value.remote(arg)) == arg
+#
+#    # Test arguments passed by ID.
+#
+#    # Test keyword arguments.
```
Comment (ericl) [inline python/ray/tests/test_mini.py]: Revert these changes?

File: python/test.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+
+@ray.remote
+def f_simple():
+    return 1
+
+
+for i in range(10):
+    print(i)
+    assert ray.get(f_simple.remote()) == 1
```
Comment (ericl) [inline python/test.py]: Remove this test file?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 3]
```diff
@@ -109,6 +126,9 @@ int64_t ClusterResourceScheduler::GetBestSchedulableNode(const TaskRequest &task
   int64_t best_node = -1;
   *total_violations = 0;
 
+  // RAY_LOG(ERROR)  << "    GET BEST NODE: " << Print();
+  // RAY_LOG(ERROR)  << "    GET BEST NODE: " << TaskRequestPrint(task_req);
+
```
Comment (ericl) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Remove these debug prints?

File: src/ray/common/scheduling/scheduling_ids.cc [hunk 4]
```diff
@@ -9,6 +9,17 @@ int64_t StringIdMap::Get(const std::string &string_id) {
   }
 };
 
+std::string StringIdMap::Get(uint64_t id) {
+  std::string id_string;
+  auto it = int_to_string_.find(id);
+  if (it == int_to_string_.end()) {
+    id_string = std::to_string(-1);
```
Comment (ericl) [inline src/ray/common/scheduling/scheduling_ids.cc]: Consider "" instead.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 5]
```diff
@@ -152,6 +173,21 @@ int64_t ClusterResourceScheduler::GetBestSchedulableNode(const TaskRequest &task
...
+  TaskRequest task_request;
+  ResourceMapToTaskRequest(task_resources, &task_request);
+  int64_t node_id = GetBestSchedulableNode(task_request, total_violations);
+
+  std::string id_string;
+  if (node_id == -1) {
+    id_string = std::to_string(-1);
+  }
+  id_string = string_to_int_map_.Get(node_id);
+  return id_string;
```
Comment (ericl) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Can replace with `return string_to_id_map_.Get(node_id)`

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 6]
```diff
@@ -160,6 +196,7 @@ bool ClusterResourceScheduler::SubtractNodeAvailableResources(
   }
   NodeResources &resources = it->second;
 
+  // RAY_LOG(ERROR) << "   SUBTRACT RESOURCES: " << TaskRequestPrint(task_req);
```
Comment (ericl) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Please remove stray log statements, here and elsewhere.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 7]
```diff
@@ -194,3 +271,226 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+      resource_capacity.available = (int64_t)it2->second;
+    }
+    if (it->first == "CPU") {
+      node_resources->capacities[CPU] = resource_capacity;
+    } else if (it->first == "GPU") {
+      node_resources->capacities[GPU] = resource_capacity;
+    } else if (it->first == "TPU") {
+      node_resources->capacities[TPU] = resource_capacity;
+    } else if (it->first == "memory") {
+      node_resources->capacities[MEM] = resource_capacity;
```
Comment (ericl) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Would be great to keep these string constants in an array instead of hard-coded here.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 8]
```diff
@@ -194,3 +271,226 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+  for (size_t i = 0; i < task_request.custom_resources.size(); i++ ) {
+    buffer << task_request.custom_resources[i].id << ":" <<
+      "(" << task_request.custom_resources[i].req.demand << ":" <<
+      task_request.custom_resources[i].req.soft << ") ";
+  }
+  buffer << "}" << std::endl;
+  return buffer.str();
+}
+
+std::string ClusterResourceScheduler::Print(void) {
```
Comment (ericl) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]:
```suggestion
std::string ClusterResourceScheduler::DebugString() {
```

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 9]
```diff
@@ -194,3 +271,226 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+  buffer << "  node custom resources {";
+  for (auto it = node_resources.custom_resources.begin();
+    it != node_resources.custom_resources.end(); ++it) {
+    buffer << it->first << ":(" << it->second.total << ":" << it->second.available << ") ";
+  }
+  buffer << "}" << std::endl;
+  return buffer.str();
+}
+
+std::string ClusterResourceScheduler::TaskRequestPrint(const TaskRequest& task_request) {
```
Comment (ericl) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Can this be a method of TaskRequest?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 10]
```diff
@@ -194,3 +271,226 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+      return false;
+    }
+    if (it1->second.available != it2->second.available) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::string ClusterResourceScheduler::NodeResourcesPrint(const NodeResources& node_resources) {
```
Comment (ericl) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Can this be a method of NodeResources?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 11]
```diff
@@ -194,3 +271,226 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+    if (itr != it->second.custom_resources.end()) {
+      string_to_int_map_.Remove(resource_id);
+      it->second.custom_resources.erase(itr);
+    }
+  }
+}
+
+
+bool ClusterResourceScheduler::EqualNodeResources(
+    const NodeResources &node_resources1, const NodeResources &node_resources2) {
```
Comment (ericl) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Can this be a method of NodeResources?

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 12]
```diff
@@ -132,31 +150,71 @@ class ClusterResourceScheduler {
...
+  void ResourceMapToNodeResources(
+      const std::unordered_map<std::string, double> &resource_map_total,
+      const std::unordered_map<std::string, double> &resource_map_available,
+      NodeResources *node_resources);
+
+  void UpdateResourceCapacity(const std::string& client_id,
+      const std::string& resource_name, int64_t resource_total);
+
+  void DeleteResource(const std::string& client_id_string,
+      const std::string& resource_name);
```
Comment (ericl) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: Document these

File: src/ray/common/scheduling/scheduling_ids.h [hunk 13]
```diff
@@ -23,6 +23,7 @@ class StringIdMap {
   /// \param String ID.
   /// \return The integer ID associated with the given string ID.
   int64_t Get(const std::string &string_id);
+  std::string Get(uint64_t id);
```
Comment (ericl) [inline src/ray/common/scheduling/scheduling_ids.h]: Document this.

File: src/ray/gcs/redis_async_context.cc [hunk 14]
```diff
@@ -56,6 +56,9 @@ Status RedisAsyncContext::RedisAsyncCommand(redisCallbackFn *fn, void *privdata,
   {
     // `redisvAsyncCommand` will mutate `redis_async_context_`, use a lock to protect it.
     std::lock_guard<std::mutex> lock(mutex_);
+    if(!redis_async_context_) {
+      return Status::NotImplemented("...");
```
Comment (ericl) [inline src/ray/gcs/redis_async_context.cc]: What's this for?

File: src/ray/raylet/format/node_manager.fbs [hunk 15]
```diff
@@ -12,31 +12,31 @@ enum MessageType:int {
...
-  FetchOrReconstruct,
+  FetchOrReconstruct,//8
   // For a worker that was blocked on some object(s), tell the raylet
   // that the worker is now unblocked. This is sent from a worker to a raylet.
-  NotifyUnblocked,
+  NotifyUnblocked,//9
   // Notify the current worker is blocked. This is only used by direct task calls;
   // otherwise the block command is piggybacked on other messages.
-  NotifyDirectCallTaskBlocked,
+  NotifyDirectCallTaskBlocked,//10
```
Comment (ericl) [inline src/ray/raylet/format/node_manager.fbs]: Please remove these debug comments.

File: src/ray/raylet/node_manager.cc [hunk 16]
```diff
@@ -14,6 +14,8 @@
 
 namespace {
 
+const bool USE_NEW_SCHEDULER = true;
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc]: Consider adding this to `ray_config_def.h` and referencing it as `initial_config.use_new_scheduler` in node manager.cc

File: src/ray/raylet/node_manager.cc [hunk 17]
```diff
@@ -1869,11 +2022,17 @@ void NodeManager::SubmitTask(const Task &task, const Lineage &uncommitted_lineag
...
     return;  // The worker may have died or is no longer processing the task.
   }
   auto const cpu_resource_ids = worker->ReleaseTaskCpuResources();
   local_available_resources_.Release(cpu_resource_ids);
+  if (USE_NEW_SCHEDULER) {
+    // TODO (ion): replace this hard coded # of CPUs.
+    std::unordered_map<std::string, double> task_request;
+    task_request.emplace("CPU", 1.);
+    new_resource_scheduler_->AddNodeAvailableResources( client_id_.Binary(), task_request);
+  }
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc]: Can we move this block above the !worker check above and return, to cleanly separate the two code paths?

File: src/ray/raylet/node_manager.cc [hunk 18]
```diff
@@ -1992,6 +2158,10 @@ void NodeManager::HandleTaskUnblocked(
         worker->AcquireTaskCpuResources(resource_ids);
         cluster_resource_map_[gcs_client_->client_table().GetLocalClientId()].Acquire(
             cpu_resources);
+        if (USE_NEW_SCHEDULER) {
+          new_resource_scheduler_->SubtractNodeAvailableResources(
+              client_id_.Binary(), cpu_resources.GetResourceMap());
+        }
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc]: This code path should never be hit in direct call mode, so we can remove this.

File: src/ray/raylet/node_manager.cc [hunk 19]
```diff
@@ -1927,10 +2090,13 @@ void NodeManager::HandleTaskBlocked(const std::shared_ptr<LocalClientConnection>
       // Release the CPU resources.
       auto const cpu_resource_ids = worker->ReleaseTaskCpuResources();
       local_available_resources_.Release(cpu_resource_ids);
+      if (USE_NEW_SCHEDULER) {
+        new_resource_scheduler_->AddNodeAvailableResources(
+            client_id_.Binary(), cpu_resource_ids.ToResourceSet().GetResourceMap());
+      }
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc]: This code path should never be hit in direct call mode, so we can remove this.

File: src/ray/raylet/node_manager.h [hunk 20]
```diff
@@ -540,6 +542,12 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
   /// unable to schedule new tasks or actors at all.
   void WarnResourceDeadlock();
 
+  /// Move tasks from runnable -> running as workers become available.
+  void DispatchDirectCallTasks();
```
Comment (ericl) [inline src/ray/raylet/node_manager.h]: Consider `DispatchScheduledTasksToWorkers()`, and updating the comment.

File: src/ray/raylet/node_manager.h [hunk 21]
```diff
@@ -540,6 +542,12 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
...
+
+  /// Move tasks from pending -> runnable queue as resources become available.
+  void NewSchedulerScheduleMoreTasks();
```
Comment (ericl) [inline src/ray/raylet/node_manager.h]: Consider `NewSchedulerSchedulePendingTasks()`, and updating the comment.

File: src/ray/raylet/node_manager.h [hunk 22]
```diff
@@ -615,6 +623,23 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
 
   /// Map of workers leased out to direct call clients.
   std::unordered_map<int, std::shared_ptr<Worker>> leased_workers_;
+
+  /// The new resource scheduler for direct task calls.
+  std::shared_ptr<ClusterResourceScheduler> new_resource_scheduler_;
+  /// Map of leased workers to their current resource usage.
+  // std::unordered_map<int, TaskRequest> leased_worker_resources_; XXX
+  std::unordered_map<int, std::unordered_map<std::string, double>>
```
Comment (ericl) [inline src/ray/raylet/node_manager.h]: Consider improving these comments.

File: src/ray/raylet/node_manager.h [hunk 23]
```diff
@@ -615,6 +623,23 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
...
+      leased_worker_resources_;
+
+  typedef std::function<void(std::shared_ptr<Worker>, ClientID spillback_to,
+                             std::string address, int port)>
+      ScheduleFn;
+
+  /// Queue of lease requests that are waiting for resources to become available.
+  /// TODO this should be a queue for each SchedulingClass
+  std::deque<std::pair<ScheduleFn, Task>> new_pending_queue_;
```
Comment (ericl) [inline src/ray/raylet/node_manager.h]: Consider renaming to `tasks_to_schedule_`.

File: src/ray/raylet/node_manager.h [hunk 24]
```diff
@@ -615,6 +623,23 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
...
+  /// Queue of lease requests that should be scheduled onto workers.
+  std::deque<std::pair<ScheduleFn, Task>> new_runnable_queue_;
```
Comment (ericl) [inline src/ray/raylet/node_manager.h]: Consider renaming to `tasks_to_dispatch_`.

File: python/ray/tests/test_mini.py (hunk 1 above)
Comment (istoica) [reply python/ray/tests/test_mini.py]: Sorry, don't remember to modify this file.

File: python/test.py (hunk 2 above)
Comment (istoica) [reply python/test.py]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 3 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: done

File: src/ray/common/scheduling/scheduling_ids.cc (hunk 4 above)
Comment (istoica) [reply src/ray/common/scheduling/scheduling_ids.cc]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 5 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: done

File: src/ray/raylet/node_manager.h (hunk 21 above)
Comment (istoica) [reply src/ray/raylet/node_manager.h]: done

File: src/ray/raylet/node_manager.h (hunk 22 above)
Comment (istoica) [reply src/ray/raylet/node_manager.h]: done

File: src/ray/raylet/node_manager.h (hunk 24 above)
Comment (istoica) [reply src/ray/raylet/node_manager.h]: done

File: src/ray/raylet/node_manager.h (hunk 23 above)
Comment (istoica) [reply src/ray/raylet/node_manager.h]: done

File: src/ray/raylet/node_manager.cc (hunk 19 above)
Comment (istoica) [reply src/ray/raylet/node_manager.cc]: done

File: src/ray/raylet/format/node_manager.fbs (hunk 15 above)
Comment (istoica) [reply src/ray/raylet/format/node_manager.fbs]: done

File: src/ray/raylet/node_manager.cc (hunk 17 above)
Comment (istoica) [reply src/ray/raylet/node_manager.cc]: done

File: src/ray/raylet/format/node_manager.fbs (hunk 15 above)
Comment (istoica) [reply src/ray/raylet/format/node_manager.fbs]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 12 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 8 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 7 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 6 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 10 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Yes, but can I do this in the next PR?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 11 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Yes, but can I do this in the next PR?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 9 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Yes, but can I do this in the next PR?

File: src/ray/common/scheduling/scheduling_ids.h (hunk 13 above)
Comment (istoica) [reply src/ray/common/scheduling/scheduling_ids.h]: done

File: src/ray/gcs/redis_async_context.cc (hunk 14 above)
Comment (istoica) [reply src/ray/gcs/redis_async_context.cc]: Ah, this was for debugging purposes. Removed it.

File: python/ray/tests/test_mini.py (hunk 1 above)
Comment (ericl) [reply python/ray/tests/test_mini.py]: Can you git checkout this file from master to undo the changes?

Comment (istoica) [reply python/ray/tests/test_mini.py]: done

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Merged, thanks!
